### PR TITLE
Creative voc decoder

### PR DESCRIPTION
**Pull request by pkrcel**
Pull request for feature in the title.

I've talked about this in this allegro.cc thead: [.VOC (Creative Voice File) samples playback with Allegro](https://www.allegro.cc/forums/thread/614933).

Actual decoding is done through [libsndfile](http://www.mega-nerd.com/libsndfile/) for which I added support in the build system.

Comments are appreciated, I'll be anyway sending a patch to the mailing list for proper procedure.

Discussion Topics:
- libsndfile can load a moltitude of formats, might consider a wider support?
- Licensing: libsndfile lives under the GNU LGPL v2.1 or v3 license [link](http://www.mega-nerd.com/libsndfile/#Licensing), this means that on systems on which allegro will not allow dynamic linking of the decoding lib (all but Windows?) it would force the GNU GPL on the whole work.
- Allegro 4 already supported Creative Voice...have I done wrong in not porting the decoder itself? (I couldn't find a guideline about this)

### CODE REVIEW

**Comment by pkrcel** [comment]
As per Beoran's observations in the linked thread, I've pushed a different implementation moving out libsndfile to a DIFFERENT file (which can be included or not) and created (I wouldn't say ported from A4) a new internal decoder for VOC files.

**Comment by pkrcel** [comment]
REmoved sndfile support from current branch, only internal decoder is active.

Still needs polishing, I guess.

File: cmake/FindGDIPLUS.cmake [hunk 1]
```diff
@@ -17,7 +17,7 @@ else()
     set(GDIPLUS_LOWERCASE 1 CACHE INTERNAL "Is GdiPlus.h spelt with lowercase?")
 endif()
 
-find_library(GDIPLUS_LIBRARY NAMES gdiplus)
+find_library(GDIPLUS_LIBRARY NAMES libgdiplus gdiplus)
```
Comment (SiegeLord) [inline cmake/FindGDIPLUS.cmake]: What's this all about?

File: addons/acodec/voc.c [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+   uint8_t x = 0;
+
+   uint16_t  timeconstant = 0;
+   uint16_t format = 0;    // can be 16-bit in Blocktype 9 (voc version > 1.20
+   uint16_t vocversion = 0;
+   uint16_t checkver = 0;  // must be  1's complement of vocversion + 0x1234
+
+   uint32_t blocklength = 0; //length is stored in 3 bites LE, gd byteorder.
+
+   if (!fp){
```
Comment (SiegeLord) [inline addons/acodec/voc.c]: The style is to add a space between parenthesis and the brace.

File: addons/acodec/acodec.h [hunk 3]
```diff
@@ -49,4 +58,9 @@ ALLEGRO_AUDIO_STREAM *_al_load_ogg_vorbis_audio_stream_f(ALLEGRO_FILE* file,
    size_t buffer_count, unsigned int samples);
 #endif
 
+#ifdef ALLEGRO_CFG_ACODEC_SNDFILE
```
Comment (SiegeLord) [inline addons/acodec/acodec.h]: These aren't used anymore, are they?

File: addons/acodec/voc.c [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+   int            samples;     /* # of samples. size = samples * sample_size */
+};
+
+/*
+ * Fills in a VOC data header with information obtained  by opening an
+ * caller-provided ALLEGRO_FILE.
+ * The datapos index will be the first data byte of the first data block which
+ * contains ACTUAL data.
+ */
+static AL_VOC_DATA *voc_open(ALLEGRO_FILE *fp){
```
Comment (SiegeLord) [inline addons/acodec/voc.c]: Brace on a new line.

File: addons/acodec/voc.c [hunk 5]
```diff
@@ -0,0 +1,342 @@
+/*
+ * Allegro5 Creative Voice Audio Reader.
```
Comment (SiegeLord) [inline addons/acodec/voc.c]: Probably could use our nice ALLEGRO ascii art here ;).

**Comment by SiegeLord** [comment]
Is there a (very small) .voc file we could add to the examples/data to make sure there's something to test this on?

File: cmake/FindGDIPLUS.cmake (hunk 1 above)
Comment (pkrcel) [inline cmake/FindGDIPLUS.cmake]: In MSYS2 environment, I had cmake which did not find gdiplus. Adding a suggested name libgdiplus fixed the problem

**Comment by pkrcel** [comment]
They are extremely small, but the ones I am working on are (most probably) covered by copyrights, since I am working with Master of Orion 2 samples.

I GUESS I could search for some free samples or conver the ones we have with SoX.

I'll look after it when I'm back from off-site duties (give me a couple days or whenabouts).

File: addons/acodec/acodec.h (hunk 3 above)
Comment (pkrcel) [inline addons/acodec/acodec.h]: These should not be in this pull request, yet another crumble of the previous implementation. Sorry ^__^

File: addons/acodec/voc.c (hunk 5 above)
Comment (pkrcel) [inline addons/acodec/voc.c]: I'll add it in a separate commit but seems that all around the Addon section of code the "Allegro" ASCII art is mostly (if not ubiquitously) absent. ^__^", that's why I havent put it there.

**Comment by SiegeLord** [comment]
I looked a bit more closely at this. I think it looks pretty good, I can fix the rest of the minor style issues myself. One thing that you should fix is that you never check the return value of `al_fread`, there's no guarantee that it will actually read the amount of bytes you request.

**Comment by pkrcel** [comment]
Will check the al_read functions and add return value control when applicable.

But could you pls let me know what are the style problems still present? That way I won't need to look at the diff post-work and I'll be able to contribute more properly to allegro next time.

Cheers
-pkrcel

File: addons/acodec/voc.c [hunk 6]
```diff
@@ -0,0 +1,362 @@
...
+   uint16_t format = 0;    // can be 16-bit in Blocktype 9 (voc version > 1.20
+   uint16_t vocversion = 0;
+   uint16_t checkver = 0;  // must be  1's complement of vocversion + 0x1234
+
+   uint32_t blocklength = 0; //length is stored in 3 bites LE, gd byteorder.
+
+   if (!fp) {
+      ALLEGRO_WARN("voc_open: Failed opening ALLEGRO_FILE");
+      return NULL;
+   };
```
Comment (SiegeLord) [inline addons/acodec/voc.c]: Extra semicolon.

File: addons/acodec/voc.c [hunk 7]
```diff
@@ -0,0 +1,362 @@
...
+   }
+
+
+static AL_VOC_DATA *voc_open(ALLEGRO_FILE *fp)
+{
+   AL_VOC_DATA *vocdata;
+   char hdrbuf[0x16];
+   size_t readcount = 0;
+
+   uint8_t  blocktype = 0;
```
Comment (SiegeLord) [inline addons/acodec/voc.c]: Extra space after uint8_t (and below as well).

File: addons/acodec/voc.c [hunk 8]
```diff
@@ -0,0 +1,362 @@
...
+          */
+         READNBYTES(fp, blocktype, 1, NULL);
+         if (blocktype != 1) {
+            ALLEGRO_WARN("voc_open: Blocktype following type 8 is not 1");
+            return NULL;
+         }
+         READNBYTES(fp, blocklength, 2, NULL);
+         READNBYTES(fp, x, 1, NULL);
+         blocklength += x<<16;
+         blocklength -=2;
```
Comment (SiegeLord) [inline addons/acodec/voc.c]: Missing space before the 2.

File: addons/acodec/voc.c [hunk 9]
```diff
@@ -0,0 +1,362 @@
...
+   if (!buffer) {
+      return NULL;
+   }
+   /*
+    * We now need to iterate over data blocks till either we hit end of file
+    * or we find a terminator block.
+    */
+   bytestoread = vocdata->samples * vocdata->sample_size;
+   while(!endofvoc && !al_feof(vocdata->file)) {
+      uint32_t blocktype =0;
```
Comment (SiegeLord) [inline addons/acodec/voc.c]: Another missing space.

**Comment by pkrcel** [comment]
Hope to have got all of them, I looked into the while file several times, but it's just like when you try to proofread your own essay for the billionth time....my eyes just keep crossing ^__^"

**Comment by SiegeLord** [comment]
Merged as 00a42a36a8147df256b8c46de9fecedbfefcd1d6 and 1d85b70342b8ddcd8f4b714e67d3e8eeef977b1b. Thanks a lot!

**Pull request closed without merging**